composer/datasets/streaming/dataset.py

```diff
@@ -0,0 +1,379 @@
+            part_max_id (int): Maximum sample ID of this partition.
+        """
+        shards = list(shards)
+        if self.shuffle:
+            np.random.shuffle(shards)
+        for shard in shards:
+            basename = get_shard_basename(shard)
+            self._download_if_missing(basename)
+            shards = shard,
+            self._load_shards(shards, part_min_id, part_max_id)
+        self._done_loading()
+
+    def _load(self) -> None:
+        """Load shards."""
+        # We find out num workers, and therefore num partitions, when __iter__ is called.
+        # From the partition, derive our shard overlap range and exact sample range.
+        world = World.from_env()
+        todo_shards, part_min_id, part_max_id = self.index.get_partition(world, self.batch_size)
+
+        # Preload all of our shards that are already available in the cache.
+        todo_shards = self._load_shards_if_downloaded(todo_shards, part_min_id, part_max_id)
+
+        # Start downloading our missing shards in a background thread, if there are any.
+        if todo_shards:
+            thread = Thread(target=self._download_thread, args=(todo_shards, part_min_id, part_max_id), daemon=True)
+            thread.start()
+        else:
+            self._done_loading()
+
+    def __len__(self) -> int:
+        """Get the length of the dataset.
+
+        Returns:
+            int: Dataset length.
+        """
+        return math.ceil(self.index.total_samples / dist.get_world_size())
+
+    def _unpack_sample(self, data: bytes) -> Dict[str, Any]:
+        """Unpack a sample dict from raw bytes.
+
+        First unpacks the str to raw bytes dict, then unpacks each field's raw bytes.
+
+        Args:
+            data (bytes): The packed bytes of the sample.
+
+        Returns:
+            Dict[str, Any]: The sample dict.
+        """
+        key_to_raw = bytes_to_sample_dict(data, self.index.fields)
+        obj = {}
+        for key, decode in self.decoders.items():
+            value = key_to_raw[key]
+            if decode:
+                value = decode(value)
+            obj[key] = value
+        return obj
+
+    def __getitem__(self, idx: int) -> Any:
+        """Get the sample at the index, assuming its shard is loaded.
+
+        Do not call this directly unless all shards have been loaded. Will crash
+        if the shard is not loaded.
+
+        Args:
+            idx (int): Sample ID.
+
+        Returns:
+            Any: The sample.
+        """
+        shard = self.index.sample_shards[idx]
+        offset = self.index.sample_shard_offsets[idx]
+        size = self.index.bytes_per_sample[idx]
+
+        basename = get_shard_basename(shard)
+        shard_filename = os.path.join(self.local, basename)
+        fp = open(shard_filename, 'rb', 0)
+        fp.seek(offset)
+        data = fp.read(size)
+        fp.close()
+
+        return self._unpack_sample(data)
+
+    def _new_growing_epoch(self) -> int:
```

Comment:
Nit: Can the function be named to be more descriptive? (E.g. `_make_new_growing_epoch`?)